Add feature flags for Phase 2

talldan opened the pull request:
## Description
An attempt at introducing feature flags for phase 2 (see #11016).

The proposal is to use webpack's define plugin (https://webpack.js.org/plugins/define-plugin/) to replace instances of `process.env.GUTENBERG_PHASE` with an integer denoting the phase (`1` or `2`) during the build.

Within the codebase, the `GUTENBERG_PHASE` variable would be used to conditionally avoid executing code:
```js
// file A
function futurePhaseTwoFeature() {
  // implementation
}

export const phaseTwoFeature = process.env.GUTENBERG_PHASE === 2 ? futurePhaseTwoFeature : undefined;

// file B
import { phaseTwoFeature } from 'file-a';

// ...
if ( process.env.GUTENBERG_PHASE === 2 ) {
  phaseTwoFeature();
}
```
When building for core, the majority code would be stripped out by webpack's minfication/dead code elimination process. 

This PR also introduces docs and linting rules to ensure that `process.env.GUTENBERG_PHASE` is used in the correct way.

talldan committed:
Add new package for editor configuration, initially containing just feature flags

Rework build commands to use correct NODE_ENV for feature flags

noisysocks commented:
Thanks for getting started on this! Nothing moves the conversation forward like putting up some strawman code 😀

> It'd be great to consider this a request for collaboration, I'm not completely familiar with the release processes, so maybe not the perfect person to deliver a finalised version of this.

@youknowriad and @gziolo should be able to give us some direction here once they're back from their meetup. Paging @WordPress/gutenberg-core, too.

> Introduce a new 'editor-configuration' package. It has a README that includes a usage example.

We probably want to give this a more generic name as `editor-configuration` implies it's only useful for the editor. In Phase 2 we'll be using Gutenberg components into other parts of WP Admin. 

> The new package contains three configuration files (development.js, plugin.js, and production.js) 

My first reaction is that overloading `NODE_ENV` with a meaning that isn't seen elsewhere in the JavaScript ecosystem will cause confusion and bite us later.

Perhaps we could have two environment variables, e.g. `NODE_ENV=production|development` and `GUTENBERG_ENV=plugin|core`?

Or perhaps we ought to ditch what Calypso does and go back to something like [Riad first suggested](https://github.com/WordPress/gutenberg/issues/11016), e.g. `NODE_ENV=production|development` and `GUTENBERG_PHASE=1|2`?

> Dead code elimination (i.e. not shipping code for disabled features). I looked into getting this working. Firstly, I don't think the technique here is works in terms of static analysis because the config file is determined at runtime

What do we need for static analysis to work? How complex can an expression be before Webpack gives up on trying to eliminate it as dead code?

> It also seems like webpack may require the sideEffects: false flag in the package.json for our packages

What's involved in getting `sideEffects` added to all our packages?

aduth commented:
> > Introduce a new 'editor-configuration' package. It has a README that includes a usage example.
> 
> We probably want to give this a more generic name as `editor-configuration` implies it's only useful for the editor. In Phase 2 we'll be using Gutenberg components into other parts of WP Admin.

This was my first thought as well. I see no reason this needs to be specific to editor at all. I think we could come up with a convention around naming specific features to allow some subgrouping (`editor/some-awesome-feature`), but it seems unnecessary at the module level.

noisysocks commented:
> A hypothetical second goal could be to have certain features be left to incubate in development and skip plugin releases until they are ready.

I don't think it's necessary (yet) but it's worth noting that `GUTENBERG_PHASE` could easily accomodate it by having `GUTENBERG_PHASE=3` for very experimental features. Then, we configure:

- `.npm run publish` to compile the scripts with `GUTENBERG_PHASE=1` 
- `./bin/build-package-zip.sh` to compile the scripts with `GUTENBERG_PHASE=2`
- `npm run dev` to compile the scripts with `GUTENBERG_PHASE=3`

Or for _super experimental_ features one could even gate the feature to `GUTENBERG_PHASE=4` where it can only be seen when manually running `GUTENBERG_PHASE=4 npm run dev`.

All food for thought. My only point is that the idea scales fairly well 🙂

noisysocks commented:
Expanding on https://github.com/WordPress/gutenberg/pull/13324#issuecomment-456800635, I thought I'd lay out a proposal for how the release process using `GUTENBERG_PHASE` would look:

- The Gutenberg `master` branch is the single source of truth and contains the code that ends up both in Gutenberg plugin releases (via `./bin/build-plugin-zip.sh`) and WordPress Core releases (via `npm run publish`).
- We gate new features that should appear in the Gutenberg plugin and *not* WordPress Core using the `GUTENBERG_PHASE` environment variable, e.g.
```js
if ( process.env.GUTENBERG_PHASE >= 2) {
	// The RSS block is under development, so only load it in the Gutenberg plugin and not WP Core
	registerBlockType( 'core/rss', rssSettings );
}
```
- We configure Webpack to inject the `GUTENBERG_PHASE` constant into our scripts at build time and to eliminate dead code that is unreachable because of a `GUTENBERG_PHASE` check.
- When building packages for WordPress Core (`npm run publish`), we set `GUTENBERG_PHASE=1`.
- When building the Gutenberg plugin zip (`./bin/build-plugin-zip.sh`), we set `GUTENBERG_PHASE=2`.

Considerations: 

- Will have to configure Travis CI to test both `GUTENBERG_PHASE=1` and `GUTENBERG_PHASE=2`.
- Will have to educate developers that, unless it's gated, code merged to `master` *will* land in the next WordPress release.

talldan committed:
Revert "Rework build commands to use correct NODE_ENV for feature flags"

This reverts commit 4cb0a39282d57abb8cd5c398ae20b9466d67b056.

Revert "Add new package for editor configuration, initially containing just feature flags"

This reverts commit 0c21fc215b4e6f0d4b415aabb9f8297635b7d148.

talldan committed:
Switch to using webpack define plugin to inject a global GUTENBERG_PHASE variable

mcsf commented:
> Thoughts on this approach?

Thanks, I feel good about this. Loose thoughts:

- For other environments and consumers, we either require the presence of `GUTENBERG_PHASE` (how?) or consistently assume `1` to be the default value whenever the flag is tested.
- The latter means being strict about `

mcsf commented:
Ouch, ignore my closing this issue. My keyboard magic with Vimium went a little too far. 😅

mcsf commented:
Thanks, this looks a lot nicer.

> Using the define plugin does add a restriction that other consumers of packages would also have to use it (or a similar solution). For example, core's webpack implementation could hard-code this value to `1`.

I'm not sure we would have a good way to enforce the presence of the env var for third-party consumers: since they can use any config for the bundling, we can't require anything in webpack, and since they can build and consume packages independently, we can't very reliably throw runtime errors if the var is missing.

It seems safer to just assume `1` is the default value. Is it worth adding an ESLint rule for any reads of the variable? For instance, requiring that any reads need to be part of a strict equality check for a value greater than 1. Examples:

*Valid:*

```js
if ( process.env.GUTENBERG_PHASE === 2 ) {
```

```js
if (
    process.env.GUTENBERG_PHASE === 2 ||
    process.env.GUTENBERG_PHASE === 3
) {
```

*Invalid:*

```js
if ( process.env.GUTENBERG_PHASE === 1 ) {
```

```js
if ( process.env.GUTENBERG_PHASE ) {
```

```js
if ( process.env.GUTENBERG_PHASE > 1 ) {
```

noisysocks commented:
> Using the define plugin does add a restriction that other consumers of packages would also have to use it (or a similar solution). For example, core's webpack implementation could hard-code this value to 1.

@talldan and I were discussing this yesterday and we concluded that it's arguably fine that we publish code on npm that requires `webpack.DefinePlugin` since this is [exactly what React does](https://unpkg.com/react@16.7.0/index.js), for example. We can communicate this requirement through [documentation similar to React's](https://reactjs.org/docs/optimizing-performance.html#webpack).

The alternative is that we modify `npm run build:packages` to use Babel plugins that define `process.env.GUTENBERG_PHASE` and eliminate dead code. This means that third parties are completely unable to use Phase 2 code.

> It seems safer to just assume 1 is the default value. Is it worth adding an ESLint rule for any reads of the variable? For instance, requiring that any reads need to be part of a strict equality check for a value greater than 1.

That's an interesting idea. I _think_ I'm into it. At the very least it makes sense to forbid `if ( GUTENBERG_PHASE === 1 )`.

One consideration: What will it look like when we "ship" Phase 2 to WordPress Core? Would we simply remove all of the `if ( GUTENBERG_PHASE === 2 )` checks?

noisysocks commented:
Assuming `1` and using ESLint to prevent `GUTENBERG_PHASE === 1` checks doesn't prevent a `ReferenceError` from being thrown when `GUTENBERG_PHASE` is referenced. 

Some off-the-bat ideas:

1) We could _require_ that third parties using WordPress packages use `webpack.DefinePlugin` to set `GUTENBERG_PHASE`. Pros: No effort on our part. Cons: Difficult for third parties.

2) Have `npm run build:scripts` prepend `if ( typeof GUTENBERG_PHASE === 'undefined' ) { GUTENBERG_PHASE = 1; }` to all of the package JavaScript files that we generate. Pros: No effort for third parties. Cons: Fairly heavy handed, probably negatively affects bundle size and parse time.

3) Have ESLint enforce that our feature gates look like this:  
    ```js
    if ( typeof GUTENBERG_PHASE !== 'undefined' && GUTENBERG_PHASE === 2) {
    ```
    Pros: No effort for third parties. Cons: We have to put up with fairly verbose checks.

4) Implement a Babel plugin which transforms feature gates that look like this:
    ```is
    if ( isGutenbergPhase( 2 ) ) {
    ```
    Into code that looks like this:
    ```js
    if ( typeof GUTENBERG_PHASE !== 'undefined' && GUTENBERG_PHASE === 2 ) {
    ```
    Pros: No effort for third parties. Cons: A lot of ✨ magic ✨.

I'm leaning towards (3).

talldan commented:
With the define plugin we can explicitly replace only `window.GUTENBERG_PHASE`, and then use linting as enforcement about the usage:
```
new DefinePlugin( {
    'window.GUTENBERG_PHASE': JSON.stringify( process.env.GUTENBERG_PHASE || 1 ),
} ),
```
☝️ that will only replace instances of `window.GUTENBERG_PHASE`, which would avoid any errors being thrown for those without the variable.

> Implement a Babel plugin which transforms feature gates that look like this:
> 
> ```if ( isGutenbergPhase( 2 ) ) {```
> 

This is a bit crazy, but it's also possible to do something like this with the define plugin:
```
		new DefinePlugin( {
			'window.isGutenbergPhase': `( ( phaseToCompare ) => phaseToCompare === ${ process.env.GUTENBERG_PHASE || 1 } )`,
		} ),
```
`if ( window.isGutenbergPhase( 2 ) )` is replaced with an immediately invoked function. Dead code elimination still seems to work as well (I imagine the minifier unwraps the function, then eliminates it). That would mean less need for linting, but has the downside that it's harder to implement for third parties. It might not be possible at all with some non-webpack bundlers.

talldan commented:
Hey @mcsf thanks for the comments. Linting is a good idea, it would also mean there's less chance of dead code elimination being bypassed if the usage is enforced.

Is there a preference for `process.env.GUTENBERG_PHASE` over `window.GUTENBERG_PHASE`?

I did notice we already have a line of code that references `process.env.NODE_ENV`, so it seem there's an assumption that `process.env` is provided by whatever build process is bundling the code:
https://github.com/WordPress/gutenberg/blob/ab0acc399ea3457833b2f7f50fd9a3bff90a57d0/packages/components/src/tooltip/index.js#L94

talldan committed:
Iterate: use window.GUTENBERG_PHASE to avoid thrown errors from an undefined global

talldan committed:
Add custom eslint rule for usage of GUTENBERG_PHASE

talldan commented:
In b7d2a5ecc7908601afff92c1073c85a05e78300d I've had a first attempt at adding a custom linting rule to govern the use of the `GUTENBERG_PHASE` variable.

I think the remaining steps:
- Add some initial documentation.
- Make sure each of the different types of build (e.g. test) have the right definition of `GUTENBERG_PHASE`.
- Code review and amendments.
- Implement a real use-case as mentioned in https://github.com/WordPress/gutenberg/pull/13511#issuecomment-458480618.
- Implement in core.

I'll try to tick some of these off over the next few days.

noisysocks commented:
> Implement a real use-case as mentioned in #13511 (comment).

The RSS, Tag Cloud, and Search blocks would be good use cases I think. (I think that only the RSS block is merged right now.)

talldan committed:
Disable new eslint rule when used in webpack config

talldan committed:
Add readme

talldan committed:
Include phase 2 features in e2e tests

talldan committed:
Allow use of GUTENBERG_PHASE in a ternary and update documentation.

talldan committed:
Add links to docs

talldan committed:
Minor docs changes

talldan commented:
I'm feeling pretty confident about this now. I've written some initial docs here detailing the proposed usage:
https://github.com/WordPress/gutenberg/blob/bbdd7a70389652f483a64f66b078f992fc62f917/docs/designers-developers/developers/feature-flags.md

This is probably a good stage to receive some initial code review feedback, and I'll try implementing one of the proposed features with feature flags, perhaps in a separate PR that integrates the two?

mkaz commented on docs/designers-developers/developers/feature-flags.md:
```diff
@@ -0,0 +1,114 @@
+
+### How it works
+
+More precisely, the `GUTENBERG_PHASE` variable isn't actually set to a value. Instead, any instances of `window.GUTENBERG_PHASE` in the codebase will be replaced during the webpack build by webpack's define plugin (https://webpack.js.org/plugins/define-plugin/).
```
Simplify explanation:

```suggestion
During the webpack build, any instances of `window.GUTENBERG_PHASE` will be replaced using webpack's define plugin (https://webpack.js.org/plugins/define-plugin/).
```

mkaz reviewed (commented):
Documentation page looks good. :+1:  

I added a minor suggestion for clarity, but overall explained the flag nicely and how to use it.

aduth commented:
> Is there a preference for `process.env.GUTENBERG_PHASE` over `window.GUTENBERG_PHASE`?
> 
> I did notice we already have a line of code that references `process.env.NODE_ENV`, so it seem there's an assumption that `process.env` is provided by whatever build process is bundling the code:

My thought here is that we already have some precedent ([another example](https://github.com/WordPress/gutenberg/blob/23a72978bbd1c856f58e2e09d359b44b8f8c1f7a/packages/blocks/src/api/validation.js#L262)), so I might weigh it in favor of following the precedent, even if just for consistency's sake.

aduth commented on docs/designers-developers/developers/feature-flags.md:
```diff
@@ -0,0 +1,114 @@
+
+The technique for handling this is known as a 'feature flag'. 
+
+## Introducing `window.GUTENBERG_PHASE`
```
While it's quite thorough, I think most people coming to this document care most about the **Basic Use** section, so at the least I might suggest moving it to precede this section, if not simplify / cut parts of this section.

aduth commented on docs/designers-developers/developers/feature-flags.md:
```diff
@@ -0,0 +1,114 @@
+
+#### Since `GUTENBERG_PHASE` is a global object, why do I have to access it through the `window` object, e.g. `window.GUTENBERG_PHASE`?
+
+Third-party consumers of WordPress npm packages may have their own build process, resulting in `window.GUTENBERG_PHASE` not being replaced at build time with the appropriate number. In JavaScript, when attempting to access an undefined global variable an error is thrown, a situation that would cause lots of issues for third-parties (or additional build configuration). Accessing the `GUTENBERG_PHA ...
```
This also hints at another problem though: They might have their own build process, so `window` can't be assumed to be defined either.

talldan replied:
Hmm, hadn't considered that. I suppose that would be usage outside the browser. 🤔 

It's a shame that the condition might end up becoming quite verbose. I'd wanted to keep it as sort and simple as possible.

aduth commented on webpack.config.js:
```diff
@@ -105,6 +106,11 @@ const config = {
 		],
 	},
 	plugins: [
+		new DefinePlugin( {
```
It surprises me that we don't already have a `DefinePlugin` plugin, and makes me worry that in places where we are expecting `NODE_ENV` substitution to incur dead code removal, that it's not actually happening. I wonder if this was changed at some point...

aduth replied:
Hmmm #755

aduth replied:
Hmmm! https://github.com/WordPress/gutenberg/pull/5267/files#diff-11e9f7f953edc64ba14b0cc350ae7b9dL216 @youknowriad

talldan replied:
@aduth I think it might have been intentional, it looks as though it happens automatically since webpack 4 whenever `mode` is specified—probably as part of their transition to requiring less config:
https://webpack.js.org/guides/production/#specify-the-mode

> Since webpack v4, specifying mode automatically configures DefinePlugin for you

I wonder whether that's still the case when `DefinePlugin` is also used manually 🤔

aduth replied:
Aha! I was looking for something exactly like that, as I'd been curious if the reason for removing it was that it was automated in such a way. Nice find.

I checked to see that it was folded correctly in what we distribute in core currently:

https://github.com/WordPress/WordPress/blob/c4505be5343f88f0dd5a0c51ab7271f1aaf46c05/wp-includes/js/dist/blocks.js#L8133

aduth commented:
Related: https://webpack.js.org/plugins/environment-plugin/

talldan committed:
Switch from window.GUTENBERG_PHASE to process.env.GUTENBERG_PHASE

talldan committed:
Update docs for feature flags. Move `Basic Use` section higher up, and simplify a sentence

talldan committed:
Ensure GUTENBERG_PHASE environment variable is available for webpack

talldan committed:
Ensure GUTENBERG_PHASE is a number

talldan committed:
Ensure GUTENBERG_PHASE is set in unit tests

talldan commented on test/unit/setup/gutenberg-phase.js:
```diff
@@ -0,0 +1,5 @@
+process.env = {
```
@gziolo I struggled to find a way to inject the `GUTENBERG_PHASE` environment variable for unit tests, I couldn't handle it in the npm script like I do with webpack. I came up with this, would be great to get your thoughts on it. Maybe you're aware of a better way.

gziolo replied:
That's very strange. I'm personally fine with such workaround. As long as it works 🤷‍♂️

talldan replied:
Unusually `process.env.GUTENBERG_PHASE = 2` doesn't seem to work. I wish I had a good explanation.

gziolo replied:
Can we add
```js
// Inject the `GUTENBERG_PHASE` global, used for feature flagging.
```
also here?

talldan replied:
👍 Done.

talldan commented:
I think I've addressed the feedback so far:
- Tidied up the docs
- Switched to `process.env.GUTENBERG_PHASE`

I've also started implementing feature flags for the search block, which unearthed a few issues - https://github.com/WordPress/gutenberg/pull/13829

This is the commit that shows how I feature flagged the block:
https://github.com/WordPress/gutenberg/pull/13829/commits/d196ea8b6e51e3567eb5de5998fc6327bca8e710

It feels like the technique used will really depend on the code that's being feature flagged.

gziolo commented on test/unit/jest.config.json:
```diff
@@ -6,7 +6,8 @@
 	"setupFiles": [
-		"core-js/fn/symbol/async-iterator"
+		"core-js/fn/symbol/async-iterator",
+		"./test/unit/setup/gutenberg-phase.js"
```
I would put do it once using [globalSetup](https://jestjs.io/docs/en/configuration#globalsetup-string). `setupFiles` runs for every test suite file and it might be too late if we ever reference the code which depends on the flag before.

By the way, the safest way to reference a file is to use `<rootDir>/test/unit/setup/gutenberg-phase.js` - which probably is the same in this case :)

gziolo replied:
`cross-env GUTENBERG_PHASE=2 wp-scripts test-unit-js --config test/unit/jest.config.json` didn't work?

talldan replied:
> `cross-env GUTENBERG_PHASE=2 wp-scripts test-unit-js --config test/unit/jest.config.json` didn't work?

No, not sure why. 

> I would put do it once using [globalSetup](https://jestjs.io/docs/en/configuration#globalsetup-string). `setupFiles` runs for every test suite file and it might be too late if we ever reference the code which depends on the flag before.

Thanks, I'll give that a go, it'll mean switching our config from json to js, assuming that's ok.

gziolo replied:
`globalSetup` is like `setupFiles` but I think it takes string - path to file. So I don't think you need to change the format of config.

talldan replied:
Oh right. Though I noticed this in the docs:
> Note: Any global variables that are defined through globalSetup can only be read in globalTeardown. You cannot retrieve globals defined here in your test suites.

So it seems as thought it wouldn't be usable for this purpose.

gziolo replied:
Now, I see that. I got confused by the example I saw in the past:

https://jestjs.io/docs/en/puppeteer#custom-example-without-jest-puppeteer-preset

gziolo replied:
> It's also worth noting that `setupFiles` will execute before `setupFilesAfterEnv` (`setupTestFrameworkScriptFile` in our version of Jest).

It should be okey as is then if we can't setup it once for all files. Those might be also two seperate run contexts.

gziolo reviewed (commented):
I didn't read all comments as they are wy too many of them already :)

My question is what will happen with the code published to npm. If I consume the code from npm I will have `process.env.GUTENBERG_PHASE` in the codebase. If I don't set this flag myself, it will be undefined. Does it mean that all code guarded with the flag will be considered a dead code unless explicitly enabled?

talldan commented on packages/eslint-plugin/configs/custom.js:
```diff
@@ -6,6 +6,7 @@ module.exports = {
 		'@wordpress/dependency-group': 'error',
 		'@wordpress/no-unused-vars-before-return': 'error',
 		'@wordpress/valid-sprintf': 'error',
+		'@wordpress/gutenberg-phase': 'error',
```
👍 I'll change that

talldan committed:
Use <rootDir> in jest config

talldan commented:
> My question is what will happen with the code published to npm. If I consume the code from npm I will have `process.env.GUTENBERG_PHASE` in the codebase. If I don't set this flag myself, it will be undefined. Does it mean that all code guarded with the flag will be considered a dead code unless explicitly enabled?

Hey @gziolo - yep, that's exactly it. We do already have some usage of `process.env.NODE_ENV` in the codebase so it's already assumed that consumers of packages will be providing the `process.env` object, though not sure if it's documented anywhere.

gziolo reviewed (commented):
One idea worth exploring would be to use config section inside `package.json` file as described in https://docs.npmjs.com/files/package.json#config (it isn't an extensive description). It gets translated into `process.env.package_config_*`. I saw a video tutorial one day on egghead.io (https://egghead.io/courses/how-to-use-npm-scripts-as-your-build-tool), that's how I know it works :)

That way we could limit the number of places which requires to redefine the same config and at the same time read the value from the JSON file if env variable is not reachable.

talldan committed:
switch to using package.json config to define the value of GUTENBERG_PHASE

talldan commented on webpack.config.js:
```diff
@@ -105,6 +106,11 @@ const config = {
+		new DefinePlugin( {
+			// Inject the `GUTENBERG_PHASE` global, used for feature flagging.
+			// eslint-disable-next-line @wordpress/gutenberg-phase
+			'process.env.GUTENBERG_PHASE': JSON.stringify( parseInt( process.env.GUTENBERG_PHASE, 10 ) || 1 ),
```
I've added a basic e2e test for the search block in https://github.com/WordPress/gutenberg/pull/13829—there don't seem to be any issues there.

talldan commented:
> One idea worth exploring would be to use config section inside `package.json` file as described in https://docs.npmjs.com/files/package.json#config (it isn't an extensive description). It gets translated into `process.env.package_config_*`. I saw a video tutorial one day on egghead.io (https://egghead.io/courses/how-to-use-npm-scripts-as-your-build-tool), that's how I know it works :)

> That way we could limit the number of places which requires to redefine the same config and at the same time read the value from the JSON file if env variable is not reachable.

Great suggestion, added in f651cb3.

gziolo commented:
Do we want to code a similar solution for CSS files? :)

gziolo reviewed (approved):
I left some minor comments to look at but in general, this looks great. I'm happy with the proposal as is. Let's make sure that all the feedback is fully addressed and let's move forward.

@aduth, would you mind doing a sanity check for ESLint rule introduced? In my testing it works as advertised, it is covered with tests and follows the same approach as other rules but I'm still not sure about the API usage. I need to write my own rule one day 😅

talldan committed:
Sort custom lint rules alphabetically

talldan committed:
Add comment about GUTENBERG_PHASE

talldan committed:
Update jest config for GUTENBERG_PHASE

talldan commented on webpack.config.js:
```diff
@@ -1,6 +1,7 @@
 /**
  * External dependencies
  */
+const { DefinePlugin } = require( 'webpack' );
```
Done 👍

talldan commented on test/unit/jest.config.json:
```diff
@@ -6,7 +6,8 @@
 	"setupFiles": [
-		"core-js/fn/symbol/async-iterator"
+		"core-js/fn/symbol/async-iterator",
+		"<rootDir>/test/unit/setup/gutenberg-phase.js"
```
I've done that, though I haven't had a chance to test it yet in the other PR. Will do that a bit later 👍

gziolo replied:
I tested it with unit tests yesterday, and it was skipping the test guarded with the phase check. I would expect the same from e2e tests.

talldan committed:
Add webpack as dependency to main package.json

gziolo commented:
I have just merged #13876 which updates tons of lines in `package-lock.json` to fix some vulnerabilities so this PR will require some update. All applied changes look great. Let's get it merged until Monday. I'm moving it to 5.1 milestone. 🎉

talldan commented:
@gziolo Thanks for the reviews, really appreciate the help 😄 

The merge conflict should now be fixed. 👍 

The only thing I'd say is that it'd be good to do some testing with how this works downstream (checking that everything works as expected when feature flagged code is integrated in core, or when it's consumed by plugins). That could always come after this is merged.

gziolo commented:
> The only thing I'd say is that it'd be good to do some testing with how this works downstream (checking that everything works as expected when feature flagged code is integrated in core, or when it's consumed by plugins). That could always come after this is merged.

Let's merge and test once published to npm. In the worst case scenario, we will publish patch later but I  personally feel pretty confident about the proposed approach.

🚢

gziolo commented:
I shared some details about this PR in #core-js channel on WordPress Slack (link requires registration at https://make.wordpress.org/chat/):
https://wordpress.slack.com/archives/C5UNMSU4R/p1550231607150600

aduth commented on webpack.config.js:
```diff
@@ -105,6 +106,11 @@ const config = {
+		new DefinePlugin( {
+			// Inject the `GUTENBERG_PHASE` global, used for feature flagging.
+			// eslint-disable-next-line @wordpress/gutenberg-phase
+			'process.env.GUTENBERG_PHASE': JSON.stringify( parseInt( process.env.npm_package_config_GUTENBERG_PHASE, 10 ) || 1 ),
```
`JSON.stringify` is often used with `DefinePlugin` because the substitution is _verbatim_, and thus for something like a string, you'd want to be certain you're replacing it with a string, not an identifier token (the difference between `foo( bar );` and `foo( "bar" )`). For a number, though, the verbatim substitution is fine, i.e. there's no difference with or without `JSON.stringify` (it's `foo( 10 )` with or without `JSON.stringify`).

aduth commented on webpack.config.js:
```diff
@@ -105,6 +106,11 @@ const config = {
 	plugins: [
+		new DefinePlugin( {
+			// Inject the `GUTENBERG_PHASE` global, used for feature flagging.
+			// eslint-disable-next-line @wordpress/gutenberg-phase
```
There's no error if this line is removed?
